### DISCUSSION TITLE
Trouble Integrating gperftools with Moose in Conda Environment

### DISCUSSION THREAD

**Discussion by hityyds** [idaholab/moose#26490] [Q&A General] [answered]
I'm using a moose environment installed with conda. Recently, I've been attempting to analyze my moose program using gperftools. I followed the instructions in the tutorial at https://mooseframework.inl.gov/application_development/profiling.html, but I'm unsure whether I need to make modifications to libmesh and wasp. My focus is specifically on using gperftools when METHOD=oprof.

> libMesh automatically adds -fno-omit-frame-pointer to METHOD=oprof builds. However, if you want to do profiling with other methods (e.g. opt, which will not give line number information), both libMesh and MOOSE should be built with CXXFLAGS=-fno-omit-frame-pointer. (On a related node, MOOSE will error if the user attempts to compile with either METHOD=devel or METHOD=dbg and with a non-empty GPERF_DIR as those methods add assertions that will make the resulting profiles misleading). When building libMesh, ensure that the METHODS environment variable is either empty (in which case the script will build opt, oprof, devel, and dbg methods), or that METHODS contains oprof (or not if you wish to profile with another method and have specified CXXFLAGS=-fno-omit-frame-pointer).

Upon updating only Petsc, I encountered some errors like this:
`/home/zu/anaconda3/envs/moose-gperf/bin/../lib/gcc/x86_64-conda-linux-gnu/10.4.0/../../../../x86_64-conda-linux-gnu/bin/ld: /home/zu/ProgramFiles/gperftools/installed/lib/libtcmalloc_and_profiler.so: undefined reference to `pthread_key_create@GLIBC_2.34'`

When attempting to update libmesh, I received the following message:
```
No LSB modules are available.
ERROR: LIBMESH_DIR=/home/zu/anaconda3/envs/moose-gperf/libmesh exists within the conda base /home/zu/anaconda3
Exiting in order to stop you from overwriting your conda environment
```
Can anyone provide guidance on how to successfully integrate gperftools? Thank you in advance for your assistance.

**Comment by GiudGiud** [comment]
Hello

If you want to hack a little and use the conda environment AND re-install a custom libmesh, then unset LIBMESH_DIR before doing so. 
Have you really optimized to the point that oprof is no longer sufficient for detecting the slower parts of the program?

Guillaume

**Reply by hityyds** [reply]
In fact, I am not familiar with oprof and gperftools at all. I just hope to analyze the memory usage and came across this tutorial. So, does moose inherently support a tool called oprof for memory analysis?

**Reply by GiudGiud** [reply, marked as answer]
OPROF is the build method which is recommended for profiling

**Reply by hityyds** [reply]
Thank you once again for your assistance.

**Reply by milljm** [reply]
I am a little confused on _hacking_ libMesh. We ship `moose-libmesh` with the following methods built: `opt, dbg, devel, oprof`. So it should not be necessary rebuilding libMesh in `OPROF` mode. But maybe I am missing something in this discussion?

```bash
❯ conda activate moose
❯ ls $LIBMESH_DIR/lib | grep oprof
libmesh_oprof.0.dylib
libmesh_oprof.dylib
libtimpi_oprof.11.dylib
libtimpi_oprof.dylib
```

**Reply by GiudGiud** [reply]
He wants to build in opt with the frame flag added

**Reply by milljm** [reply]
Ah, I think is see the disconnect... I figured those instructions would explain that we have performance tools from google available as a Conda package:

```bash
conda activate moose
conda install moose-pprof
```
(I receive a bunch of warnings during installation on my Apple Silicon machine, but resulting `pprof` still seems to work fine)


After which, you will need to re-activate the environment in order to have GPERF_DIR set:
```bash
conda deactivate
conda activate moose
echo $GPERF_DIR
/Users/milljm/mamba3/envs/moose
```
